### PR TITLE
Zperf improvements tcp

### PR DESCRIPTION
**Pull request by luca-fancellu**
This serie is improving zperf library code quality and it's implementing multiple TCP connections handling in the TCP receiver module.

### CODE REVIEW

File: subsys/net/lib/zperf/zperf_tcp_receiver.c [hunk 1]
```diff
@@ -184,34 +178,35 @@ static void tcp_server_session(void)
...
 		} else {
 		use_existing_ipv6:
 			/* Use existing IP */
-			in6_addr = zperf_get_default_if_in6_addr();
-			if (!in6_addr) {
-				NET_ERR("Unable to get IPv4 by default");
+			const struct in6_addr *addr =
+				zperf_get_default_if_in6_addr();
+			if (!addr) {
+				NET_ERR("Unable to get IPv6 by default\n");
```
Comment (rlubos) [inline subsys/net/lib/zperf/zperf_tcp_receiver.c:223]: Why adding `\n` though? It'll print an extra empty line in the log.

File: subsys/net/lib/zperf/zperf_tcp_receiver.c [hunk 2]
```diff
@@ -128,34 +123,35 @@ static void tcp_server_session(void)
...
 		} else {
 		use_existing_ipv4:
 			/* Use existing IP */
-			in4_addr = zperf_get_default_if_in4_addr();
-			if (!in4_addr) {
-				NET_ERR("Unable to get IPv4 by default");
+			const struct in_addr *addr =
+				zperf_get_default_if_in4_addr();
+			if (!addr) {
+				NET_ERR("Unable to get IPv4 by default\n");
```
Comment (rlubos) [inline subsys/net/lib/zperf/zperf_tcp_receiver.c:175]: Ditto, `\n` does not seem needed.

File: subsys/net/lib/zperf/zperf_tcp_receiver.c [hunk 3]
```diff
@@ -100,10 +98,50 @@ static void tcp_received(const struct sockaddr *addr, size_t datalen)
...
+
+	if (address->sa_family == AF_INET) {
+		port = ntohs(net_sin(address)->sin_port);
+	} else {
+		port = ntohs(net_sin6(address)->sin6_port);
+	}
+
+	ret = zsock_bind(pollfd->fd, address, sizeof(*address));
+	if (ret < 0) {
+		NET_ERR("Cannot bind IPv%d TCP port %d (%d)\n",
```
Comment (rlubos) [inline subsys/net/lib/zperf/zperf_tcp_receiver.c:115]: Please fix other log messages as well, as I see there are more places like this - `\n` is not needed with the logger, as it adds new line implicitly.

File: subsys/net/lib/zperf/zperf_tcp_receiver.c [hunk 4]
```diff
@@ -254,52 +274,62 @@ static void tcp_server_session(void)
 			}
 
 			switch (i) {
-			case SOCK_ID_IPV4_LISTEN:
-				addr = &addr_ipv4;
-				__fallthrough;
-			case SOCK_ID_IPV6_LISTEN:{
-				int sock = zsock_accept(fds[i].fd, addr, &addrlen);
+			case SOCK_ID_IPV4_LISTEN ... SOCK_ID_IPV6_LISTEN: {
```
Comment (rlubos) [inline subsys/net/lib/zperf/zperf_tcp_receiver.c:277]: I'm actually not sure if we should use GNU C extensions, AFAIK we should strive for portability and not rely on specific compiler features. @carlescufi Is that correct, or am I exaggerating?

**Review by rlubos — COMMENTED** [review]
@luca-fancellu Thank you for taking our previous comments into consideration. It looks pretty good now.

I'm just wondering though - we now can have mutliple `zperf` download sessions running at the same time, but only one session callback. This is fine, however I'm thinking that perhaps it'd make sense to extend the callback structure with some unique session identifier? Right now, the application have no way to link events with a specific session. Socket FD could for instance be used as a session ID, as the system guarantees its uniqueness. WDYT?

**Review by rlubos — COMMENTED** [review]
Nevermind, wrong button...

File: subsys/net/lib/zperf/zperf_tcp_receiver.c (hunk 1 above)
Comment (luca-fancellu) [reply subsys/net/lib/zperf/zperf_tcp_receiver.c:223]: Sure I'll remove `\n` from all the logs

File: subsys/net/lib/zperf/zperf_tcp_receiver.c (hunk 4 above)
Comment (luca-fancellu) [reply subsys/net/lib/zperf/zperf_tcp_receiver.c:277]: This is a valid point, however also clang supports case ranges, and I've seen some place in the codebase where they are used.
I don't have, however, a strong objection to remove them, in the end case(s) are just if(s).

Let me know your preference

**Comment by luca-fancellu** [comment]
> I'm just wondering though - we now can have mutliple `zperf` download sessions running at the same time, but only one session callback. This is fine, however I'm thinking that perhaps it'd make sense to extend the callback structure with some unique session identifier? Right now, the application have no way to link events with a specific session. Socket FD could for instance be used as a session ID, as the system guarantees its uniqueness. WDYT?

I think would be nice, but I've checked also the UDP part and even if it's handling multiple connections, it is not giving any info about the zperf session.
In the case someone is going to implement this improvement, it should be a different PR in my opinion, the current work is aligned between TCP and UDP callbacks.
Please, let me know your thoughts about that.

File: subsys/net/lib/zperf/zperf_tcp_receiver.c (hunk 4 above)
Comment (jukkar) [reply subsys/net/lib/zperf/zperf_tcp_receiver.c:277]: IMHO these extensions should be avoided unless there is no other alternative. Here we have only couple of values to check so I would only use standard C in this case.

File: subsys/net/lib/zperf/zperf_tcp_receiver.c [hunk 5]
```diff
@@ -259,52 +274,62 @@ static void tcp_server_session(void)
...
+				       addrlen);
+
 				break;
 			}
 
-			case SOCK_ID_IPV4_DATA:
-				addr = &addr_ipv4;
-				__fallthrough;
-			case SOCK_ID_IPV6_DATA:
+			case (SOCK_ID_IPV6_LISTEN + 1) ... (SOCK_ID_MAX - 1):
```
Comment (jukkar) [inline subsys/net/lib/zperf/zperf_tcp_receiver.c:314]: same thing here

Comment (luca-fancellu) [reply subsys/net/lib/zperf/zperf_tcp_receiver.c:314]: ok, I will change that

File: subsys/net/lib/zperf/zperf_tcp_receiver.c (hunk 1 above)
Comment (luca-fancellu) [reply subsys/net/lib/zperf/zperf_tcp_receiver.c]: done

File: subsys/net/lib/zperf/zperf_tcp_receiver.c (hunk 4 above)
Comment (luca-fancellu) [reply subsys/net/lib/zperf/zperf_tcp_receiver.c]: fixed

File: subsys/net/lib/zperf/zperf_tcp_receiver.c (hunk 5 above)
Comment (luca-fancellu) [reply subsys/net/lib/zperf/zperf_tcp_receiver.c]: fixed

**Review by jukkar — APPROVED** [review]
LGTM

**Review by rlubos — APPROVED** [review]

**Pull request merged**